Add AWS S3 storage backend

dobesv opened the pull request:
Add support for cache storage to AWS S3.

dobesv committed:
Add AWS S3 storage backend

ecraig12345 commented on packages/cache/src/S3CacheStorage.ts:
```diff
@@ -0,0 +1,172 @@
+ * input has not started providing data after a given time after
+ * its creation.
+ */
+class TimeoutStream extends Transform {
```
Can you move these helper streams to a shared file instead of copying them from `AzureBlobCacheStorage.ts`?

dobesv replied:
Done

ecraig12345 reviewed (changes requested):
Thanks for this PR, but most of the current users of lage/backfill won't want a tree of AWS dependencies pulled in by default.

An alternative would be to specify the AWS deps as `peerDependencies` that are marked as optional in `peerDependenciesMeta` (and update the provider to throw if the dep isn't present). The peer deps would need to be specified in both the `backfill-cache` and `backfill` packages.

Also, due to the potentially-missing dependency and the way lage's types rollup works, it would be best to avoid directly references the types in `backfill-config`--so `clientConfig?: object` in `S3CacheStorageOptions`.

dobesv commented:
- I moved the shared Azure/S3 classes to their own files so they can be shared between those two backends.
- I changed the types in `config` so they don't need to reference the aws-sdk package types

ecraig12345 commented:
Sorry for the extremely delayed response again. We just merged the `backfill` repo into the `lage` repo so it's easier to manage, so I'm about to archive this repo and its PRs.

Regarding this specific PR: I appreciate the contribution and recognize that it would likely be useful for others, but since our team has limited resources to devote to `lage`/`backfill`, I'm hesitant to take on maintenance of another backend that we don't use and wouldn't be able to test. As a bit of context, `lage`, `backfill`, `beachball`, and `just` (and a few others) were made by people from some teams in Office--so as much as we'd like to take a more robust, community-oriented approach to making great open source tools for both our teams and others, it's never been our primary responsibility and makes it hard to find the right balance.

As an alternative, I realized recently that since most users of `lage` don't use remote caching at all, probably a better approach is to handle *all* heavy backends including Azure as plugins (https://github.com/microsoft/lage/issues/1037). Basically modifying the existing `CustomStorageConfig` so it takes a module name or path. At that point you could publish the S3 implementation as a separate npm package, and I'd be happy to add references to that in the relevant docs.
